it/multipart/src/test/java/com/linecorp/armeria/common/multipart/BodyParts.java

```diff
@@ -0,0 +1,112 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the Licenses
+ */
+
+package com.linecorp.armeria.common.multipart;
+
+import static java.util.Objects.requireNonNull;
+
+import java.nio.file.OpenOption;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Function;
+
+import com.linecorp.armeria.common.CommonPools;
+import com.linecorp.armeria.common.QueryParams;
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.common.stream.StreamMessages;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.util.concurrent.EventExecutor;
+
+/**
+ * A utility class for saving upload files and getting query parameters from {@link Multipart}.
+ */
+@UnstableApi
+public class BodyParts {
+    /**
+     *  Immutable HTTP Multipart result.
+     */
+    public static class CollectedBodyParts {
+        private final QueryParams queryParams;
+        private final Map<String, List<Path>> files;
+
+        CollectedBodyParts(QueryParams queryParams, Map<String, List<Path>> files) {
+            this.queryParams = queryParams;
+            this.files = files;
+        }
+
+        /**
+         * Return whole {@link QueryParams} from the Multipart.
+         */
+        public QueryParams getQueryParams() {
+            return queryParams;
+        }
+
+        /**
+         * Returns a {@link Map} whose key is the name from {@link BodyPart} and value is {@link List} of
+         * {@link Path} where to find uploaded files.
+         */
+        public Map<String, List<Path>> getFiles() {
+            return files;
+        }
+    }
+
+    /**
+     * Collect the query params and write the uploaded file to the given {@link Path} with
+     * {@link OpenOption} from the specified {@link Multipart}.
+     * If the {@link OpenOption} is not specified, defaults to {@link StandardOpenOption#CREATE},
+     * {@link StandardOpenOption#TRUNCATE_EXISTING} and {@link StandardOpenOption#WRITE}.
+     *
+     * @param multipart the source of {@link Multipart} to be processed
+     * @param mappingFileName the {@link Function} that generates a {@link Path} from the name in the part
+     * @param options the {@link OpenOption} specifying how the file is opened
+     * @return a {@link CompletableFuture} that completes with the aggregated {@link CollectedBodyParts} or an
+     *         error
+     *
+     * @see StreamMessages
+     */
+    public static CompletableFuture<CollectedBodyParts> collect(
```

Comment:
Are we going to do this in the follow-up PR?

Reply:
Not sure, do u think it's better to add to this PR? this will add new method to existing Multipart and will user confuse the behavior with aggregate?

Reply:
Also the 
```
 public static CompletableFuture<CollectedBodyParts> collect(
            Multipart multipart,
            Function<String, Path> mappingFileName,
            OpenOption... options) {
```
is already write to file or aggregate to memory. But here 
```
interface Multipart {
   CompletableFuture<List<T>> collect(Function<? super BodyPart, CompletableFuture<? extends T>> function);
}
```
is asking user do to by self, right? It means we won't need `ContentAwareMultipartCollector` to do `bodyPart.aggregateWithPooledObjects` and `bodyPart.writeFile` and just delegate onNext to user?

Reply:
So we probably don't need this as a public API. 😄

Reply:
Let me think ^^;;; No strong opinion here yet. Just curious if we add it, it's hard to change. But if we decide to add, I can add in this PR or in next PR.

Reply:
I think it will still be useful to add `Multipart.collect(): CompletableFuture` for users who want to write a test case that consumes `Multipart`, just like we have `HttpRequest.collect()`. Please feel free to refactor in this or follow-up PR. :bow: